validator/storage_location.rego

```diff
@@ -0,0 +1,59 @@
+#
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+package templates.gcp.GCPStorageLocationConstraintV1
+
+import data.validator.gcp.lib as lib
+
+deny[{
+	"msg": message,
+	"details": metadata,
+}] {
+	constraint := input.constraint
+	lib.get_constraint_params(constraint, params)
+
+	asset := input.asset
+	asset.asset_type == "storage.googleapis.com/Bucket"
+
+	# Check if resource is in exempt list
+	exempt_list := params.exemptions
+	matches := {asset.name} & cast_set(exempt_list)
+	count(matches) == 0
+
+	# Check that location is in allowlist/denylist
+	target_locations := params.locations
+	asset_location := asset.resource.data.location
+	location_matches := ({upper(asset_location)} & cast_set(target_locations)) | ({lower(asset_location)} & cast_set(target_locations))
```

Comment:
I'd like the match to be case insensitive because CAI outputs have locations in upper case while [documentation](https://cloud.google.com/storage/docs/locations) and many terraform templates in practice have locations in lower case.
Not sure whether we want to advertise because this is probably not the neatest way to perform case insensitive search (e.g. it does not cover cases like 'Us-central1') but I have not found a better way from https://www.openpolicyagent.org/docs/latest/language-reference/.